### ISSUE TITLE
Add discussion of only including grouped or aggregated fields in SELECT

### ISSUE THREAD

**Issue by ethanwhite** [datacarpentry/sql-ecology-lesson#52]
In queries with GROUP BY it is important to only include grouped or aggregated fields in the select statements. Otherwise results are typically not what the user expects. This should be discussed in 02.

**Comment by tpoisot** [comment]
Should we add a section with potential gotchas? This is a good example, as is the `WHERE` / `HAVING` distinction.

**Comment by ethanwhite** [comment]
When I teach this I demo what goes wrong if you don't do it; so I show them the gotcha component, but I do it as part of teaching that section, not as a separate section. This also tends to keep it brief because it can be added on as a 60 second section after the last exercise: "One important thing to remember when you're using GROUP BY is..."

So, my initial inclination is to just include it at the end of GROUP BY, but I'd be happy to look at an alternative construction if you think the alternative approach is better.

**Comment by waveform80** [comment]
Might be worth pointing out that MySQL and SQLite (the latter largely for compatibility with the former) are both overly lax in this regard, and that almost every other engine (including PostgreSQL), will simply throw an error if non-aggregated clauses containing un-grouped columns (with no functional dependency on the grouped key, if one wants to be really strict) appear in the SELECT clause.

Thankfully, MySQL is *finally* [moving away from permitting this](https://dev.mysql.com/doc/refman/5.7/en/group-by-handling.html) so I'm vaguely hoping SQLite will follow suit in a year or two (or at least provide an option to disable this behaviour as MySQL did at first).

In other words, it might be worth pointing out that in SQLite this works but will give you unexpected (non-deterministic) results; in all other engines (including MySQL in the near future) this will simply bomb, so it's worth avoiding.

**Comment by ethanwhite** [comment]
> In other words, it might be worth pointing out that in SQLite this works but will give you unexpected (non-deterministic) results; in all other engines (including MySQL in the near future) this will simply bomb, so it's worth avoiding.

I understand the desire to add the 2nd part, but it feels like starting to bloat the lesson. Every little extra tidbit we give folks is potentially one less important thing they end up remembering. It also risks confusion about SQL being (basically) a standard vs. a bunch of different languages (MySQL, PostgreSQL, SQLite), which is a common point of confusion with our learners.

In this case the advice to avoid it is sound whether it errors or does something counter intuitive. If SQLite changes this behavior (and I certainly hope they do 😄) then we can swap out the discussion on non-intuitive behavior for the error.

**Comment by waveform80** [comment]
Fair enough - there's definitely something to be said for keeping lessons simple.

(personally I always refer to SQL as a "standard", with emphasized quotation marks ;)
